### PR TITLE
chore: enable eslint-plugin-perfectionist on scope-manager package

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: continues #8479
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Enables sorting on `packages/scope-manager`.

💖

### CODE REVIEW

File: packages/scope-manager/src/referencer/ClassVisitor.ts [hunk 1]
```diff
@@ -27,19 +29,70 @@ class ClassVisitor extends Visitor {
...
+
+  protected TSAbstractAccessorProperty(
+    node: TSESTree.TSAbstractAccessorProperty,
+  ): void {
+    this.visitPropertyDefinition(node);
+  }
+
+  /////////////////////
+  // Visit selectors //
+  /////////////////////
```
Comment (bradzacher) [inline packages/scope-manager/src/referencer/ClassVisitor.ts:78]: This sorting is wrong.
The `visit*` helpers used to all be grouped at the top.

I don't have a problem with them being at the end -- but the comments need to be repositioned correctly.

File: packages/scope-manager/src/referencer/PatternVisitor.ts [hunk 2]
```diff
@@ -15,6 +17,25 @@ type PatternVisitorCallback = (
...
+  readonly #assignments: (
+    | TSESTree.AssignmentExpression
+    | TSESTree.AssignmentPattern
+  )[] = [];
+
+  readonly #callback: PatternVisitorCallback;
+  readonly #restElements: TSESTree.RestElement[] = [];
+  readonly #rootPattern: TSESTree.Node;
+  public readonly rightHandNodes: TSESTree.Node[] = [];
+  constructor(
```
Comment (bradzacher) [inline packages/scope-manager/src/referencer/PatternVisitor.ts:29]:
```suggestion
  readonly #assignments: (
    | TSESTree.AssignmentExpression
    | TSESTree.AssignmentPattern
  )[] = [];
  readonly #callback: PatternVisitorCallback;
  readonly #restElements: TSESTree.RestElement[] = [];
  readonly #rootPattern: TSESTree.Node;
  public readonly rightHandNodes: TSESTree.Node[] = [];

  constructor(
```

File: packages/scope-manager/src/referencer/Referencer.ts [hunk 3]
```diff
@@ -383,6 +174,10 @@ class Referencer extends Visitor {
     this.close(node);
   }
 
+  ///////////////////
+  // Visit helpers //
+  ///////////////////
```
Comment (bradzacher) [inline packages/scope-manager/src/referencer/Referencer.ts:179]: incorrect comment placement

File: packages/scope-manager/src/referencer/Referencer.ts [hunk 4]
```diff
@@ -430,26 +224,17 @@ class Referencer extends Visitor {
     // this defines no local variables
   }
 
+  /////////////////////
+  // Visit selectors //
+  /////////////////////
```
Comment (bradzacher) [inline packages/scope-manager/src/referencer/Referencer.ts:229]: incorrect comment placement

File: packages/scope-manager/src/referencer/TypeVisitor.ts [hunk 5]
```diff
@@ -87,6 +35,10 @@ class TypeVisitor extends Visitor {
     // don't visit the property
   }
 
+  /////////////////////
+  // Visit selectors //
+  /////////////////////
```
Comment (bradzacher) [inline packages/scope-manager/src/referencer/TypeVisitor.ts:40]: incorrect comment placement

**Review by bradzacher — CHANGES_REQUESTED** [review]

**Review by bradzacher — DISMISSED** [review]

File: packages/scope-manager/src/ScopeManager.ts [hunk 6]
```diff
@@ -40,25 +41,15 @@ class ScopeManager {
   /**
    * The root scope
    */
+  readonly #options: ScopeManagerOptions;
   public globalScope: GlobalScope | null;
```
Comment (bradzacher) [inline packages/scope-manager/src/ScopeManager.ts:45]: The JSDoc comment didn't move

File: packages/scope-manager/src/ScopeManager.ts [hunk 7]
```diff
@@ -105,6 +106,16 @@ class ScopeManager {
...
+
+  protected nestScope(scope: Scope): Scope {
+    if (scope instanceof GlobalScope) {
+      assert(this.currentScope == null);
+      this.globalScope = scope;
+    }
+    this.currentScope = scope;
+    return scope;
+  }
   public acquire(node: TSESTree.Node, inner = false): Scope | null {
```
Comment (bradzacher) [inline packages/scope-manager/src/ScopeManager.ts:119]: The JSDoc comment again

File: packages/scope-manager/src/analyze.ts [hunk 8]
```diff
@@ -52,27 +59,22 @@ interface AnalyzeOptions {
...
   /**
    * The source type of the script.
    */
   sourceType?: SourceType;
-
-  // TODO - remove this in v8
-  /**
-   * @deprecated This option never did what it was intended for and will be removed in a future major release.
-   */
-  emitDecoratorMetadata?: boolean;
```
Comment (bradzacher) [inline packages/scope-manager/src/analyze.ts:64]: The todo comment moved when it shouldn't have

File: packages/scope-manager/src/referencer/Reference.ts [hunk 9]
```diff
@@ -1,13 +1,14 @@
...
 import type { Variable } from '../variable';
 
+import { createIdGenerator } from '../ID';
+
 enum ReferenceFlag {
   Read = 0x1,
-  Write = 0x2,
   ReadWrite = 0x3,
+  Write = 0x2,
 }
```
Comment (bradzacher) [inline packages/scope-manager/src/referencer/Reference.ts:12]: Value vs label ordering for enums again.

Perhaps we should disable sorting on enums?

File: packages/scope-manager/src/referencer/Reference.ts [hunk 10]
```diff
@@ -19,8 +20,8 @@ interface ReferenceImplicitGlobal {
 const generator = createIdGenerator();
 
 enum ReferenceTypeFlag {
-  Value = 0x1,
   Type = 0x2,
+  Value = 0x1,
 }
```
Comment (bradzacher) [inline packages/scope-manager/src/referencer/Reference.ts:25]: Value vs label ordering again

File: packages/scope-manager/src/referencer/Reference.ts [hunk 11]
```diff
@@ -59,10 +60,10 @@ class Reference {
    * If reference is writeable, this is the node being written to it.
    * @public
    */
-  public readonly writeExpr?: TSESTree.Node | null;
-
   public readonly maybeImplicitGlobal?: ReferenceImplicitGlobal | null;
 
+  public readonly writeExpr?: TSESTree.Node | null;
+
```
Comment (bradzacher) [inline packages/scope-manager/src/referencer/Reference.ts:66]: The JSDoc comment didn't move

File: packages/scope-manager/src/referencer/Reference.ts [hunk 12]
```diff
@@ -78,10 +79,6 @@ class Reference {
   /**
    * True if this reference can reference values
    */
-  public get isValueReference(): boolean {
-    return (this.#referenceType & ReferenceTypeFlag.Value) !== 0;
-  }
-
```
Comment (bradzacher) [inline packages/scope-manager/src/referencer/Reference.ts:84]: Jsdoc comment didn't move

File: packages/scope-manager/src/scope/ScopeBase.ts [hunk 13]
```diff
@@ -216,6 +218,61 @@ abstract class ScopeBase<
...
+
+      return true;
+    };
+
+    if (!resolve()) {
+      this.delegateToUpperScope(ref);
+    }
+  };
+
   public readonly variableScope: VariableScope;
```
Comment (bradzacher) [inline packages/scope-manager/src/scope/ScopeBase.ts:276]: Jsdoc comment didn't move

File: packages/scope-manager/src/scope/ScopeBase.ts [hunk 14]
```diff
@@ -367,38 +365,14 @@ abstract class ScopeBase<
...
-    if (variables == null) {
-      variables = [];
-      this.#declaredVariables.set(node, variables);
-    }
-    if (!variables.includes(variable)) {
-      variables.push(variable);
-    }
-  }
-
   protected defineVariable(
```
Comment (bradzacher) [inline packages/scope-manager/src/scope/ScopeBase.ts:376]: Jsdoc comment didn't move

**Review by bradzacher — CHANGES_REQUESTED** [review]

File: packages/scope-manager/src/ScopeManager.ts (hunk 6 above)
Comment (JoshuaKGoldberg) [reply packages/scope-manager/src/ScopeManager.ts:45]: Fixed by https://github.com/azat-io/eslint-plugin-perfectionist/issues/97 🚀

File: packages/scope-manager/src/lib/lib.ts
[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/scope-manager/src/lib/lib.ts:1]: Good call, they don't. I'll add these to `ignores`.

File: package.json [hunk 15]
```diff
@@ -99,7 +99,7 @@
     "eslint-plugin-jest": "^27.9.0",
     "eslint-plugin-jsdoc": "^47.0.2",
     "eslint-plugin-jsx-a11y": "^6.8.0",
-    "eslint-plugin-perfectionist": "^3.2.0",
+    "eslint-plugin-perfectionist": "^3.8.0",
```
Comment (JoshuaKGoldberg) [inline package.json:102]: Also done in https://github.com/typescript-eslint/typescript-eslint/pull/9976.

**Comment by JoshuaKGoldberg** [comment]
Ok I'm tired of dealing with merge conflicts, will just merge now. Someone yell at me if anything's amiss.

Note the touchups on the ESLint config to fix `main` branch issues.

**Pull request merged**